client/src/useJob.js

```diff
@@ -0,0 +1,133 @@
+import { useFetchContext } from "./context";
+import { useState } from "react";
+import { getDerivUrl, getIngestUrl, getStatusUrl } from "./routes";
+import { getAgeString } from "./useJobHelper";
+
+const useJob = ({ category, children }) => {
+    const {
+        action: { makeRequest, getJSON },
+    } = useFetchContext();
+    const [statusText, setStatusText] = useState([]);
+    const [published, setPublished] = useState();
+    const [clickWarning, setClickWarning] = useState("");
+    const [action, setAction] = useState(null);
+    const [ingestInfo, setIngestInfo] = useState("");
+    const [clickable, setClickable] = useState(false);
+    const [updateTimer, setUpdateTimer] = useState(null);
+    const getPublishedStatusText = ({ derivatives, ingesting, published }) => {
+        if (published) {
+            if (ingesting) {
+                return ["ingesting now; cannot be edited"];
+            }
+            setAction({
+                onClick: ingest,
+                text: "ingest now",
+            });
+            return ["queued for ingestion; cannot be edited"];
+        } else if (derivatives.expected === derivatives.processed) {
+            setClickable(true);
+            return ["ready"];
+        }
+        if (!derivatives.building) {
+            setAction({
+                onClick: buildDerivatives,
+                text: "build derivatives",
+            });
+        }
+        const percentDone = 100 * (derivatives.processed / derivatives.expected);
+        setClickable(true);
+        return ["derivatives: " + percentDone.toFixed(2) + "% built"];
+    };
+
+    const getJobStatusText = ({ derivatives, documents, audio, ingesting, published }) => {
+        if (derivatives.expected === 0 && documents === 0 && audio === 0) {
+            return ["empty job"];
+        }
+        const pageCount = parseInt(derivatives.expected / 3);
```

Comment:
The magic number "3" here refers to the number of derivative types generated per page. Maybe we should define this as a constant (`derivativeTypeCount`, maybe) to make the code a little more readable.